lib/editor/actions/map/stopStrategies.js

```diff
@@ -553,3 +561,175 @@ export function removeStopFromPattern (pattern: Pattern, stop: GtfsStop, index:
+  const numberControlPoints = numberSegments - 1
+  // Add control points and segments to the list to delete and return
+  return [
+    addRangeToArray(deletedControlPoints, deletionStartIndex + 1, numberControlPoints), // Add one for immediate next control point.
+    addRangeToArray(deletedSegments, deletionStartIndex, numberSegments)
+  ]
+}
+
+/**
+ * Generates a new segment for the fromPoint and toPoint and inserts it into the new patternSegments array.
+ */
+const generateSegmentAndInsert = async (clonedPatternSegments: Array<Coordinates>, fromPoint: ?ControlPoint, toPoint: ?ControlPoint, insertionPoint: number) => {
+  if (!fromPoint || !toPoint) return
+  const points = [fromPoint.point.geometry.coordinates, toPoint.point.geometry.coordinates]
+  const segment = await getSegment(points, true)
+  if (segment && segment.coordinates) clonedPatternSegments[insertionPoint] = segment.coordinates
+}
+
+/**
+ * Converts an object with numerical indices to an array.
+ */
+const objectToArray = (obj): Array<any> => Object.keys(obj)
+  .sort((a, b) => a - b)
+  .reduce((acc, currentKey) => {
+    if (obj[currentKey] === null) return acc
+    return [...acc, obj[currentKey]]
+  }, [])
+
+/**
+ * Updates the shape_dist_traveled values for each control point after a change in the pattern has occurred.
+ */
+const updateShapeDistTraveled = (controlPoints: Array<ControlPoint>, segments: Array<Array<number>>): void => {
+  const stopCps = getStopsFromControlPoints(controlPoints)
+
+  stopCps.forEach((cp, index) => {
+    if (index === 0) controlPoints[cp.cpIndex].shapeDistTraveled = 0.0
+    else {
+      // previous cp.shape_dist_traveled + all segment distances between
+      const previousStopCP = stopCps[index - 1]
+      let newShapeDistTraveled = previousStopCP.cpIndex
+      for (let i = previousStopCP.cpIndex; i < cp.cpIndex; i++) {
+        const segmentDistance = lineDistance(lineString(segments[i]), 'meters')
+        newShapeDistTraveled += segmentDistance
+      }
+      controlPoints[cp.cpIndex].shapeDistTraveled = newShapeDistTraveled
+    }
+  })
+}
+
+/**
+ * Updates the shapes (segments and control points) after the pattern has been reordered using the pattern stop card dragging feature.
+ * This method will override several of the segments around the old location of the moved stop, and around
+ * the new location where the stop is being moved to.
+ */
+export function updateShapesAfterPatternReorder (oldPattern: Pattern, newPatternStops: Array<PatternStop>, oldPatternStopSequence: number) {
+  return async function (dispatch: dispatchFn, getState: getStateFn) {
+    const {controlPoints, patternSegments} = getControlPoints(getState())
+
+    // Use object copies of the arrays to preserve indices, we'll convert back after juggling deletions
+    // $FlowFixMe: Unfixable flow error probably a good indication this is a hack.
+    let clonedControlPoints = Object.assign({}, clone(controlPoints))
```

Comment:
I'm wrapping in an object so that I can set the indices to a value without removing them and filter the "deleted" values later but now that I'm saying that out loud I realise an array can do that too.. in conclusion, not sure, I will refactor.

Reply:
I have removed the whole arrays as objects idea, it was basically there to avoid juggling array indices after deletions. Moving back to arrays created a couple awkward sections of code but this is probably better.